source/common/upstream/upstream_impl.cc

```diff
@@ -67,9 +69,38 @@ uint64_t parseFeatures(const envoy::api::v2::Cluster& config,
       config.upstream_bind_config().freebind().value()) {
     features |= ClusterInfoImpl::Features::FREEBIND;
   }
+  if ((connection_options.has_tcp_keepalive() && !connection_options.tcp_keepalive().disable()) ||
+      (config.upstream_connection_options().has_tcp_keepalive() &&
+       !config.upstream_connection_options().tcp_keepalive().disable())) {
+    features |= ClusterInfoImpl::Features::USE_TCP_KEEPALIVE;
+  }
   return features;
 }
 
+Network::TcpKeepaliveConfig
+parseTcpKeepaliveConfig(const envoy::api::v2::Cluster& config,
+                        const envoy::api::v2::UpstreamConnectionOptions connection_options) {
+  if (connection_options.has_tcp_keepalive()) {
+    return Network::TcpKeepaliveConfig{
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_probes,
+                                        absl::optional<uint32_t>()),
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_time,
+                                        absl::optional<uint32_t>()),
+        PROTOBUF_GET_WRAPPED_OR_DEFAULT(connection_options.tcp_keepalive(), keepalive_interval,
+                                        absl::optional<uint32_t>())};
+  } else if (config.upstream_connection_options().has_tcp_keepalive()) {
```

Comment:
These two cases can be refactored and collapsed into one, just operating on a different object.